File: javascript/anagram/anagram.js
function Anagram(word){

  this.word = word;

  this.matches = function(){
    
    const array = [].concat(...arguments);

    var lower = this.word.toLowerCase();

    for (var k = 0; k < array.length; k++){
      if (lower === array[k].toLowerCase() || lower.split('').sort().join('') !== array[k].toLowerCase().split('').sort().join('')){
        array.splice(k,1);
        k--;
      } 
    }
    return array;

  }

}

module.exports = Anagram;

File: ruby/anagram/anagram.rb
class Anagram
	attr_accessor :base, :sorted_base

	def initialize(word)
		@base = word.downcase
	end


	def match(words)
		words.select(&method(:is_anagram?))
	end

	private

	def rebased_word(word)
		word.downcase.split('').sort.join
	end

	def sorted_base
		@sorted_base ||= rebased_word(base)
	end

	def is_anagram?(word)
		return false if matches_base?(word)
		sorted_base == rebased_word(word)
	end

	def matches_base?(word)
		base == word.downcase
	end
end